Design.md

```diff
@@ -136,46 +121,88 @@ This API will update the existing storage tenant with the provided `cd_tenant_id
+* If the tenant account credentials were not found in `assumeRoleCache`, the `AssumeRoleBackbeat` API must be called as superadmin and added to the `assumeRoleCache` cache.
+* If the `AssumeRoleBackbeat` API returns a `Role does not exist` error, then invoke the [`SetupAssumeRole` subroutine](#SetupAssumeRole-subroutine)).
+* `assumeRoleCache` cache will be a **[key, value]** pair of **[Role_Arn, Temporary_Credentials]** respectively.
+    * Each `assumeRoleCache` entry will have a 50-minute TTL (its session token is valid only for 60 minutes).
+    * When accessing credentials on the cache, refresh the credentials as well as `needsRefresh` (using AWS SDK `refresh()`), as they are subject to expiration.
+    * Cache design can be found [here](#Cache-Design).
+    
 ### Create User
 This API creates a user on Vault.
 1. `create-user` api will be called using assumed role credentials.
-    * `cd_user_id` is stored as username.
+    * The tenant user's `cdUserId` is stored as `username` in the corresponding Vault user.
+    * The tenant user's `username` and `role` enum value are stored in the Vault user path as `<tenantUsername>::<roleEnumValue>`.
 1. `attach-policy` for the new user will be called using the assumed role credentials with full S3 and IAM access permissions.
 1. `create-s3credentials` api will be called using assumed role credentials
 
 ### Create User Example Activity Diagram
 ![Create User](Design_Files/OSE-OSIS-Create-User.png)
 
+### List Users
+This API will list users on Vault.
+1. The `list-users` API can be called with following parameters:
+    * `offset`: The start index of tenants to return (optional)
+    * `limit`: The maximum number of tenants to return (optional)
+1. `list-users` api will be called using assumed role credentials.
+    * If `offset` is present, call the `list-users` until the `offset` to identify the `marker`. Then return the `list-users` result from the `marker`
+
 ### Query Users
 This API will query users on Vault using a `filter` parameter.
 1. `list-users` api will be called using assumed role credentials.
+    * If `offset` is present, call the `list-users` until the `offset` to identify the `marker`. Then return the `list-users` result from the `marker`.
 1. Filter the result users using the `filter` parameter in the request
 
-### List Users
-This API will list users on Vault.
-1. `list-users` api will be called using assumed role credentials.
-
 ### Get User w/ userid
 This API will return the user.
-1. `get-user` api will be called using assumed role credentials with cd_user_id as username.
+1. The `get-user` API will be called using assumed role credentials, with tenant user's user ID as username.
 
 ### Get User w/ canonical ID
 This API will return the user.
 1. `get-user` api will be called using assumed role credentials with canonical_id as canonical_user_id?
```

Comment:
Is this still an open question?

Reply:
Yes we dont have any other ID on user to store for canonical id. So we can return the cduserid as `canonicalUserID` but if the create user request is made with  `canonicalUserID`, we don't know where to store it. This is still an open question and can wait till milestone 2

Reply:
How can we make a CreateUser request using canonical ID, we have to provide user ID and canonical ID is generated?
Reference: https://docs.aws.amazon.com/IAM/latest/APIReference/API_CreateUser.html

Reply:
done 569dba03652832856a3dc38329c6cbf4872fa1d4